### DISCUSSION TITLE
Form values organized by step and pre-populating related fields?

### DISCUSSION THREAD

**Discussion by lscmaro** [ivan-dalmet/formiz#77] [Q&A] [answered]
Is there a way to have Formiz separate out the form values, step by step and pre-populate related fields in another step with previous step form values? 

I have a [Codesandbox](https://codesandbox.io/s/happy-archimedes-ywjyw) where I'm trying to pre-populate but can't seem to get it. And also seem to be having issues with Radio options setting and updating correctly (Which is likely my lack of React skills)

**Comment by lscmaro** [comment]
Wow .. Organizing by step is as simple as dot notation! Love it.
`<MyField name='step_one.firstname' label='First Name' />` 
`<MyField name='step_one.lastname' label='Last Name' />`

**Reply by ivan-dalmet** [reply]
Is it all good or do you still need help? ^^

**Reply by lscmaro** [reply]
Yea, I'm still having issues with pre-populating field values. Strings work just fine for `defaultValue` , but trying to use previously entered input from the form is my issue. Radio selections are all good now so that's a plus. Thanks for making this simple, cool plugin too 👍

**Comment by ivan-dalmet** [comment, marked as answer]
@lscmaro You can handle this use case with a custom hook like this:

```jsx
import { useEffect } from "react";
import { useForm } from "@formiz/core";
import get from "lodash/get";

export const useDependentField = (fieldNameOrigin, fieldNameDestination) => {
  const { setFieldsValues, values, fields } = useForm({ subscribe: { fields: [fieldNameOrigin, fieldNameDestination] } });
  const fieldOriginValue = get(values, fieldNameOrigin);  // Needs Lodash here to handle nested fields names
  const fieldDestinationPristine = get(fields, fieldNameDestination)?.isPristine;  // Needs Lodash here to handle nested fields names

  useEffect(() => {
    // Update only if the destination field is pristine
    if (fieldDestinationPristine) {
      setFieldsValues({ [fieldNameDestination]: fieldOriginValue });
    }
  }, [
    setFieldsValues,
    fieldNameDestination,
    fieldOriginValue,
    fieldDestinationPristine
  ]);
};
```

And then you can use it like that:

```jsx
export const StepOne = (props) => {
  const { formWizard } = props;

  useDependentField("customer_info.firstname", "ship_info.firstname");
  useDependentField("customer_info.lastname", "ship_info.lastname");

  return (
    <div className={`step-one ${formWizard.currentStep ? "isActive" : ""}`}>
      <InputField
        name="customer_info.firstname"
        label="First Name"
        required="First Name is Required"
      />
      <InputField
        name="customer_info.lastname"
        label="Last Name"
        required="Last Name is Required"
      />
    </div>
  );
};
```

There is your [Codesandbox](https://codesandbox.io/s/fancy-cherry-ip1cr?file=/src/useDependentField.js) with this code 😊

https://user-images.githubusercontent.com/9749061/127139273-75e9bdba-2d07-49c3-a27c-ab90b75ebd41.mov


I hope this can help, feel free to ask for more details if it's not clear 🤗

**Reply by lscmaro** [reply]
This helps a lot. Thank you! 🥳

**Reply by lscmaro** [reply]
One issue i'm having with this approach is Required fields are no longer required and I'm trying to work through that.  Maybe you have an idea why that validation could be lost or a quick fix?

UPDATE: I tried another route since I was having issue with require validation. Probably would be better to use the form hooks and such though, but this seemed to work.

I created some helper methods to basically create a new object that can be passed to `setFieldsValues`

```
  const setObjectValues = (obj, name, v, k) => (
      obj[`${name}.${k}`] = v
  );
  
  export const objectMap = (useObj, name, setObj) => Object.fromEntries(
      Object.entries(useObj).map(
          ([k, v], i) => [k, setObjectValues(setObj, name, v, k, i)]
      )
  );
```

Then I just passed the values after the step was submitted.

```
const newObj = {};
objectMap(formWizard.values?.customer_info, 'shipping_info', newObj);
formWizard.setFieldsValues(newObj);
```

**Reply by ivan-dalmet** [reply]
I'm not sure to understand the problem or the solution ^^
But if it's works :) 
Feel free to ask for more help if you're facing other issues.

**Reply by lscmaro** [reply]
Not that it matters now but, for better understanding ... Adding `useDependentField` code  makes the fields on 1st step not required.  You can click 'continue' button without adding values to 1st step required fields. These should rather trigger error when empty and continue is clicked. [CodeSandbox](https://codesandbox.io/s/fancy-cherry-ip1cr?file=/src/useDependentField.js)

**Reply by ivan-dalmet** [reply]
Oh yes that's weird 🧐

**Reply by ivan-dalmet** [reply]
Ok I've found the issue, it's an issue with `debounce` on fields and `setFieldsValues` 😕
I will need more time to look into it.

If anyone is interested, the following code is working (unless the debounce value is change on a field 😕)

```jsx
import { useEffect } from "react";
import { useForm } from "@formiz/core";
import get from "lodash/get";

export const useDependentField = (fieldNameOrigin, fieldNameDestination) => {
  const { setFieldsValues, values, fields } = useForm({
    subscribe: { fields: [fieldNameOrigin, fieldNameDestination] }
  });
  const fieldOriginValue = get(values, fieldNameOrigin); // Needs Lodash here to handle nested fields names
  const fieldDestinationPristine = get(fields, fieldNameDestination)
    ?.isPristine; // Needs Lodash here to handle nested fields names

  useEffect(() => {
    // Update only if the destination field is pristine
    if (fieldDestinationPristine) {
      const timer = setTimeout(() => {
        setFieldsValues({ [fieldNameDestination]: fieldOriginValue });
      }, 100); // Default debounce time 😕
      return () => clearTimeout(timer);
    }
  }, [
    setFieldsValues,
    fieldNameDestination,
    fieldOriginValue,
    fieldDestinationPristine
  ]);
};
```

If I've time, I will try to debug this use case when a `setFieldsValues` is called on mount.

**Reply by ivan-dalmet** [reply]
I've created an issue https://github.com/ivan-dalmet/formiz/issues/82
